File: clean-care-admin/src/services/reviewService.ts
import axios from 'axios';
import type { AxiosInstance } from 'axios';
import { API_CONFIG } from '../config/apiConfig';

/**
 * Review Service
 * Handles all review-related API calls
 */

// ==================== Types ====================

export interface Review {
    id: number;
    complaintId: number;
    userId: number;
    rating: number;
    comment: string | null;
    createdAt: string;
    user: {
        id: number;
        firstName: string;
        lastName: string;
        avatar: string | null;
    };
}

export interface ReviewSubmission {
    rating: number;
    comment?: string;
}

export interface ReviewAnalytics {
    averageRating: number;
    totalReviews: number;
    reviewPercentage: number;
    ratingDistribution: {
        1: number;
        2: number;
        3: number;
        4: number;
        5: number;
    };
    recentReviews: Array<{
        id: number;
        rating: number;
        comment: string | null;
        createdAt: string;
        complaint: {
            id: number;
            title: string;
        };
        user: {
            firstName: string;
            lastName: string;
        };
    }>;
}

export interface ReviewAnalyticsFilters {
    cityCorporationCode?: string;
    zoneId?: number;
    wardId?: number;
    startDate?: string;
    endDate?: string;
}

export interface ApiError {
    name: string;
    statusCode: number;
    message: string;
    errors?: any[];
}

// ==================== Service Class ====================

class ReviewService {
    private apiClient: AxiosInstance;

    constructor() {
        this.apiClient = axios.create({
            baseURL: API_CONFIG.BASE_URL,
            timeout: API_CONFIG.TIMEOUT,
            withCredentials: true,
            headers: {
                'Content-Type': 'application/json',
            },
        });

        this.setupInterceptors();
    }

    /**
     * Setup request and response interceptors
     */
    private setupInterceptors() {
        // Request interceptor - Add auth token
        this.apiClient.interceptors.request.use(
            (config) => {
                try {
                    const token = localStorage.getItem('accessToken');
                    if (token) {
                        config.headers.Authorization = `Bearer ${token}`;
                    }
                } catch (error) {
                    console.error('Error getting token:', error);
                }
                return config;
            },
            (error) => Promise.reject(error)
        );

        // Response interceptor - Handle errors
        this.apiClient.interceptors.response.use(
            (response) => response,
            (error) => {
                if (error.response?.status === 401) {
                    // Token expired or invalid - redirect to login
                    if (window.location.pathname !== '/login') {
                        window.location.assign('/login');
                    }
                }
                return Promise.reject(this.handleError(error));
            }
        );
    }

    /**
     * Handle API errors and convert to ApiError
     */
    private handleError(error: any): ApiError {
        if (axios.isAxiosError(error)) {
            const statusCode = error.response?.status || 500;
            const message = error.response?.data?.message || 'An error occurred';
            const errors = error.response?.data?.errors;

            return {
                name: 'ApiError',
                statusCode,
                message,
                errors,
            } as ApiError;
        }

        return {
            name: 'ApiError',
            statusCode: 500,
            message: error.message || 'An unexpected error occurred',
        } as ApiError;
    }

    /**
     * Submit a review for a complaint
     * @param complaintId - ID of the complaint to review
     * @param review - Review data (rating and optional comment)
     * @returns Created review
     */
    async submitReview(
        complaintId: number,
        review: ReviewSubmission
    ): Promise<Review> {
        try {
            const response = await this.apiClient.post<{
                success: boolean;
                data: Review;
            }>(`/api/complaints/${complaintId}/review`, review);

            return response.data.data;
        } catch (error) {
            console.error('Error submitting review:', error);
            throw error;
        }
    }

    /**
     * Get all reviews for a specific complaint
     * @param complaintId - ID of the complaint
     * @returns Array of reviews
     */
    async getComplaintReviews(complaintId: number): Promise<Review[]> {
        try {
            const response = await this.apiClient.get<{
                success: boolean;
                data: Review[];
            }>(`/api/complaints/${complaintId}/reviews`);

            return response.data.data;
        } catch (error) {
            console.error('Error fetching complaint reviews:', error);
            throw error;
        }
    }

    /**
     * Get review analytics (Admin only)
     * @param filters - Optional filters (city corporation, zone, ward, date range)
     * @returns Review analytics data
     */
    async getReviewAnalytics(
        filters: ReviewAnalyticsFilters = {}
    ): Promise<ReviewAnalytics> {
        try {
            const params: any = {};

            // Add filters to params if provided
            if (filters.cityCorporationCode) {
                params.cityCorporationCode = filters.cityCorporationCode;
            }
            if (filters.zoneId !== undefined) {
                params.zoneId = filters.zoneId;
            }
            if (filters.wardId !== undefined) {
                params.wardId = filters.wardId;
            }
            if (filters.startDate) {
                params.startDate = filters.startDate;
            }
            if (filters.endDate) {
                params.endDate = filters.endDate;
            }

            const response = await this.apiClient.get<{
                success: boolean;
                data: ReviewAnalytics;
            }>('/api/admin/complaints/analytics/reviews', { params });

            return response.data.data;
        } catch (error) {
            console.error('Error fetching review analytics:', error);
            throw error;
        }
    }
}

// Export singleton instance
export const reviewService = new ReviewService();

File: lib/services/review_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import '../config/api_config.dart';
import '../models/review_model.dart';

/// Service for managing complaint reviews
class ReviewService {
  final String baseUrl = ApiConfig.baseUrl;

  /// Submit a review for a complaint
  /// 
  /// [complaintId] - ID of the complaint to review
  /// [rating] - Rating from 1-5 stars
  /// [comment] - Optional comment (max 300 characters)
  /// 
  /// Returns the created ReviewModel object
  /// Throws Exception on error
  Future<ReviewModel> submitReview({
    required String complaintId,
    required int rating,
    String? comment,
  }) async {
    try {
      // Validate rating
      if (rating < 1 || rating > 5) {
        throw Exception('Rating must be between 1 and 5');
      }

      // Validate comment length
      if (comment != null && comment.length > 300) {
        throw Exception('Comment must be 300 characters or less');
      }

      final url = Uri.parse('$baseUrl/api/complaints/$complaintId/review');
      
      final response = await http.post(
        url,
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer ${await _getAuthToken()}',
        },
        body: jsonEncode({
          'rating': rating,
          if (comment != null && comment.isNotEmpty) 'comment': comment,
        }),
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        final data = jsonDecode(response.body);
        return ReviewModel.fromJson(data['data']);
      } else {
        final error = jsonDecode(response.body);
        throw Exception(error['message'] ?? 'Failed to submit review');
      }
    } catch (e) {
      print('❌ ReviewService.submitReview error: $e');
      rethrow;
    }
  }

  /// Get all reviews for a complaint
  /// 
  /// [complaintId] - ID of the complaint
  /// 
  /// Returns list of ReviewModel objects
  /// Throws Exception on error
  Future<List<ReviewModel>> getReviews(String complaintId) async {
    try {
      final url = Uri.parse('$baseUrl/api/complaints/$complaintId/reviews');
      
      final response = await http.get(
        url,
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer ${await _getAuthToken()}',
        },
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        final reviewsData = data['data'] as List;
        return reviewsData.map((json) => ReviewModel.fromJson(json)).toList();
      } else {
        final error = jsonDecode(response.body);
        throw Exception(error['message'] ?? 'Failed to fetch reviews');
      }
    } catch (e) {
      print('❌ ReviewService.getReviews error: $e');
      rethrow;
    }
  }

  /// Get the user's review for a specific complaint (if exists)
  /// 
  /// [complaintId] - ID of the complaint
  /// 
  /// Returns ReviewModel object or null if no review exists
  /// Throws Exception on error
  Future<ReviewModel?> getUserReview(String complaintId) async {
    try {
      final reviews = await getReviews(complaintId);
      
      // Get current user ID from auth token
      final userId = await _getCurrentUserId();
      
      // Find review by current user
      final review = reviews.firstWhere(
        (review) => review.userId.toString() == userId,
        orElse: () => throw Exception('No review found'),
      );
      return review;
    } catch (e) {
      // Return null if no review found
      if (e.toString().contains('No review found')) {
        return null;
      }
      print('❌ ReviewService.getUserReview error: $e');
      rethrow;
    }
  }

  /// Helper method to get auth token from storage
  Future<String> _getAuthToken() async {
    try {
      final sp = await SharedPreferences.getInstance();
      final token = sp.getString('accessToken');
      
      if (token == null || token.isEmpty) {
        throw Exception('No authentication token found. Please log in.');
      }
      
      return token;
    } catch (e) {
      print('❌ ReviewService._getAuthToken error: $e');
      rethrow;
    }
  }

  /// Helper method to get current user ID from auth token
  /// 
  /// Note: This method fetches the user profile from the API to get the user ID.
  /// The user ID is stored in the UserModel returned by the /api/users/me endpoint.
  Future<String> _getCurrentUserId() async {
    try {
      final url = Uri.parse('$baseUrl/api/users/me');
      
      final response = await http.get(
        url,
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer ${await _getAuthToken()}',
        },
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        final user = data['user'];
        
        if (user != null && user['id'] != null) {
          return user['id'].toString();
        } else {
          throw Exception('User ID not found in response');
        }
      } else {
        final error = jsonDecode(response.body);
        throw Exception(error['message'] ?? 'Failed to fetch user profile');
      }
    } catch (e) {
      print('❌ ReviewService._getCurrentUserId error: $e');
      rethrow;
    }
  }
}
